### PR TITLE
chore(clip): validate attention mask

### PR DESCRIPTION
**Pull request by dcvz**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -8,13 +8,19 @@ wgpu-backend = ["burn-wgpu"]
...
-burn-ndarray = "0.10.0"
-burn-tch = "0.10.0"
+burn = "0.11.0"
+burn-ndarray = "0.11.0"
+burn-tch = "0.11.0"
 clap = { version = "4.4.7", features = ["derive"] }
 serde = { version = "1.0.192", features = ["derive"] }
+
+[patch.crates-io]
+burn-wgpu = { git = "https://github.com/Tracel-AI/burn", rev = "be5bb33" }
```
Comment (dcvz) [inline Cargo.toml:23]: patching since current main has fixes + new additions we need

File: src/transformers/clip.rs [hunk 2]
```diff
@@ -118,10 +119,7 @@ impl ClipConfig {
         let token_embedding = nn::EmbeddingConfig::new(self.vocab_size, self.embed_dim).init();
         let position_embedding =
             nn::EmbeddingConfig::new(self.max_position_embeddings, self.embed_dim).init();
-        let position_ids = Tensor::arange(0..self.max_position_embeddings)
-            .unsqueeze()
-            // burn::Module does not support Int tensors yet, so we have to cast to Float
```
Comment (dcvz) [inline src/transformers/clip.rs:123]: This was fixed upstream, so took the change to update this as well

File: src/transformers/clip.rs [hunk 3]
```diff
@@ -240,37 +234,26 @@ impl<B: Backend> ClipAttention<B> {
     fn shape(&self, xs: Tensor<B, 3>, seq_len: usize, bsz: usize) -> Tensor<B, 4> {
         xs.reshape([bsz, seq_len, self.num_attention_heads, self.head_dim])
             .swap_dims(1, 2)
-        // .contiguous() // TODO: Figure out if this is needed or if we can abstract over memory
```
Comment (dcvz) [inline src/transformers/clip.rs:243]: In Burn we don't have to worry about contiguous.

Screenshot 2023-11-22 at 01 08 39

File: src/transformers/clip.rs [hunk 4]
```diff
@@ -240,37 +234,26 @@ impl<B: Backend> ClipAttention<B> {
...
-    fn forward(&self, xs: Tensor<B, 3>, causal_attention_mask: &Tensor<B, 2>) -> Tensor<B, 3> {
+    fn forward(&self, xs: Tensor<B, 3>, causal_attention_mask: Tensor<B, 4>) -> Tensor<B, 3> {
         let [bsz, seq_len, embed_dim] = xs.dims();
         let query_states = self.q_proj.forward(xs.clone()) * self.scale;
         let proj_shape = [bsz * self.num_attention_heads, seq_len, self.head_dim];
 
-        let query_states = self
-            .shape(query_states, seq_len, bsz)
-            .reshape(proj_shape)
-            .to_full_precision();
```
Comment (dcvz) [inline src/transformers/clip.rs:254]: By default, Tensors are already in Float format so I don't think we have to do this.

File: src/transformers/clip.rs [hunk 5]
```diff
@@ -240,37 +234,26 @@ impl<B: Backend> ClipAttention<B> {
...
+            .reshape(proj_shape);
         let src_len = key_states.dims()[1];
         let attn_weights = query_states.matmul(key_states.swap_dims(1, 2));
 
         let attn_weights = attn_weights
             .reshape([bsz, self.num_attention_heads, seq_len, src_len])
-            .add(
-                causal_attention_mask
-                    .to_owned()
-                    .unsqueeze::<4>()
```
Comment (dcvz) [inline src/transformers/clip.rs:271]: mask is now a Tensor of shape 4, so we don't have to expand to that shape anymore

File: src/transformers/clip.rs [hunk 6]
```diff
@@ -346,47 +327,48 @@ pub struct ClipTextTransformer<B: Backend> {
 
 impl<B: Backend> ClipTextTransformer<B> {
     // https://github.com/huggingface/transformers/blob/674f750a57431222fa2832503a108df3badf1564/src/transformers/models/clip/modeling_clip.py#L678
-    fn build_causal_attention_mask(bsz: usize, seq_len: usize, device: &B::Device) -> Tensor<B, 2> {
-        let mask = Tensor::full_device([bsz, seq_len, seq_len], f32::MIN, device);
-        let mask = zero_lower_diagonal(mask); // zero out the lower diagonal
-                                              // expand mask
-        mask.unsqueeze()
+    fn build_causal_attention_mask(bsz: usize, seq_len: usize, device: &B::Device) -> Tensor<B, 4> {
+        let mut mask_vec = Vec::with_capacity(seq_len * seq_len);
```
Comment (Trivernis) [inline src/transformers/clip.rs:331]: maybe we could move that to a utils module

File: src/transformers/clip.rs (hunk 5 above)
Comment (Trivernis) [reply src/transformers/clip.rs:271]: where did the additional dimensions come from?

**Review by Trivernis — APPROVED** [review]

File: src/transformers/clip.rs (hunk 5 above)
Comment (dcvz) [reply src/transformers/clip.rs:271]: In `build_casual_attention_mask` before it was a Tensor<B, 2> and now it's a Tensor<B, 4>

Comment (Trivernis) [reply src/transformers/clip.rs:271]: Oh, I see

File: src/transformers/clip.rs (hunk 6 above)
Comment (dcvz) [reply src/transformers/clip.rs]: I did, and will merge now!
